core/src/main/java/com/graphhopper/routing/ch/CHPreparationGraph.java

```diff
@@ -0,0 +1,870 @@
+        }
+
+        @Override
+        public int getOrigEdgeKeyFirstAB() {
+            return origEdgeKeyFirst;
+        }
+
+        @Override
+        public int getOrigEdgeKeyFirstBA() {
+            return origEdgeKeyFirst;
+        }
+
+        @Override
+        public int getOrigEdgeKeyLastAB() {
+            return origEdgeKeyLast;
+        }
+
+        @Override
+        public int getOrigEdgeKeyLastBA() {
+            return origEdgeKeyLast;
+        }
+
+        @Override
+        public String toString() {
+            return getNodeA() + "-" + getNodeB() + " (" + origEdgeKeyFirst + ", " + origEdgeKeyLast + ") " + getWeightAB();
+        }
+    }
+
+    private static class OrigGraph {
+        private final IntArrayList firstEdgesByNode;
+        private final IntArrayList adjNodes;
+        private final IntArrayList edgesAndFlags;
+
+        private OrigGraph(IntArrayList firstEdgesByNode, IntArrayList adjNodes, IntArrayList edgesAndFlags) {
+            this.firstEdgesByNode = firstEdgesByNode;
+            this.adjNodes = adjNodes;
+            this.edgesAndFlags = edgesAndFlags;
+        }
+
+        PrepareGraphOrigEdgeExplorer createOutOrigEdgeExplorer() {
+            return new OrigEdgeIteratorImpl(this, false);
+        }
+
+        PrepareGraphOrigEdgeExplorer createInOrigEdgeExplorer() {
+            return new OrigEdgeIteratorImpl(this, true);
+        }
+
+        static class Builder {
+            private final IntArrayList fromNodes = new IntArrayList();
+            private final IntArrayList toNodes = new IntArrayList();
+            private final IntArrayList edgesAndFlags = new IntArrayList();
+            private int maxFrom = -1;
+            private int maxTo = -1;
+
+            void addEdge(int from, int to, int edge, boolean fwd, boolean bwd) {
+                fromNodes.add(from);
+                toNodes.add(to);
+                edgesAndFlags.add(getEdgeWithFlags(edge, fwd, bwd));
+                maxFrom = Math.max(maxFrom, from);
+                maxTo = Math.max(maxTo, to);
+
+                fromNodes.add(to);
+                toNodes.add(from);
+                edgesAndFlags.add(getEdgeWithFlags(edge, bwd, fwd));
+                maxFrom = Math.max(maxFrom, to);
+                maxTo = Math.max(maxTo, from);
+            }
+
+            OrigGraph build() {
+                int[] sortOrder = IndirectSort.mergesort(0, fromNodes.elementsCount, new IndirectComparator.AscendingIntComparator(fromNodes.buffer));
+                sortAndTrim(fromNodes, sortOrder);
+                sortAndTrim(toNodes, sortOrder);
+                sortAndTrim(edgesAndFlags, sortOrder);
+                return new OrigGraph(buildFirstEdgesByNode(), toNodes, edgesAndFlags);
+            }
+
+            private int getEdgeWithFlags(int edge, boolean fwd, boolean bwd) {
+                // we use only 30 bits for the edge Id and store two access flags along with the same int
```

Comment:
This is a bit low. Can we avoid this reduction to 30 bits somehow?

Reply:
Oh good point, I almost forgot about this. But 30bits should be enough, because this is only for base graph edges. And if they exceeded 30bits the edge-based preparation would fail anyway, because there are more than twice as many shortcuts than base edges. This is only used for edge-based.